Update documentation (#250)
xenit-eu/dynamic-extensions-for-alfresco: documentation/Extension_Point_Content_Model.md

# Content Model

> Dynamic Extensions provides a convenient way of registering custom content models with the Alfresco Data Dictionary. (See the <a href="http://wiki.alfresco.com/wiki/Data_Dictionary_Guide">Data Dictionary Guide</a> for more info on content models.)

Reviewer: ```suggestion
Dynamic Extensions provides a convenient way of registering custom content models with the Alfresco Data Dictionary. (See the <a href="https://docs.alfresco.com/5.0/concepts/content-modeling-about.html?m=2">Data Dictionary Guide</a> for more info on content models.)
```

> Dynamic Extensions provides a convenient way of registering custom content models with the Alfresco Data Dictionary. (See the <a href="http://wiki.alfresco.com/wiki/Data_Dictionary_Guide">Data Dictionary Guide</a> for more info on content models.)

Reviewer 2: This file uses HTML rather than markdown? It renders fine on Github, so it's not really a problem, just a bit strange.

To have XML content models registered automatically, place them in this location in your extension:

> <code>/META-INF/alfresco/models</code> <a href="https://github.com/laurentvdl/example-dynamic-extension/tree/master/src/main/resources/META-INF/alfresco/models">See example project</a>

Reviewer: ```suggestion
<code>/META-INF/alfresco/models</code> <a href="https://github.com/xenit-eu/example-dynamic-extension/tree/master/src/main/resources/META-INF/alfresco/models">See example project</a>
```

When Dynamic Extensions installs an extension it scans for the presence of XML files in this folder and registers/updates the content models with the Data Dictionary. Uninstalling an extension does not remove the model as this would cause the model to disappear during restarts.

Dynamic Extensions assumes the XML files in this folder contain content models, any other XML content will yield an error. It ignores everything in this folder that is not an XML file.

## Reference

> * <a href="http://wiki.alfresco.com/wiki/Data_Dictionary_Guide">Data Dictionary Guide</a> (Alfresco Wiki)

Reviewer: ```suggestion
* <a href="https://docs.alfresco.com/5.0/concepts/content-modeling-about.html?m=2">Content modelling</a> (Alfresco Documentation)
```

> * [Content modelling](https://docs.alfresco.com/5.0/concepts/content-modeling-about.html?m=2) (Alfresco Documentation)

Reviewer: ```suggestion
* [Content modeling](https://docs.alfresco.com/5.0/concepts/content-modeling-about.html?m=2) (Alfresco Documentation)
```

Note that you cannot make breaking changes to your model. Alfresco will refuse to update the model in this case.
(same behaviour as manual upload to Data Dictionary / Models)

## Direct Model DAO access

During development, it can be usefull to bypass the registration of models via the repository when doing heavy refactoring of a model. As Alfresco does not allow breaking changes (such as adding a required aspect or removing types that are in use), you can opt for a bypass of this validation during __development__.

To do so, override the default RepositoryModelRegistrar by adding this component definition to your project.

This will cause models to be unregistered whenever your extension is stopped (or updated).

```java
@Component("modelRegistrar")
public class ModelRegistrar extends DAOModelRegistrar implements InitializingBean {
    @Autowired
    public ModelRegistrar(DictionaryDAO dictionaryDAO, M2ModelListProvider m2ModelListProvider) {
        super(dictionaryDAO, m2ModelListProvider);
    }

    @Override
    public void afterPropertiesSet() throws Exception {
        registerModels();
    }
}
```

> > Note that this can break you data consistency and thus should only be used on a local development Alfresco !

Reviewer 2: That's pretty important, maybe put it in bold rather than blockquote.
